### PR TITLE
simple support cluster and basicauth in v2

### PR DESCRIPTION
**Pull request by DoubleSpout**
1. change the `table. nkeys`, on old version use `next` instead, support openresty under v1.15.8.1
2. add simple etcd cluster support, loop chosen node, not have any error deal
3. add v2 basic auth in `new(opts)`
4. fix `timeout default 5000` second bug

btw, when shall we use etcd v3 ? ^ _ ^

### CODE REVIEW

File: lib/resty/etcd.lua [hunk 1]
```diff
@@ -282,7 +383,7 @@ local function get(self, key, attr)
     end
 
     local res, err = _request(self, "GET",
-                              self.endpoints.full_prefix .. normalize(key),
+                            choose_one_endpoint(self, self.endpoints).full_prefix .. normalize(key),
```
Comment (membphis) [inline lib/resty/etcd.lua]: bad indentation

**Comment by membphis** [comment]
`add simple etcd cluster support`, need to add new test cases.

and please sign the `CLA`, you are using a non github account in your commit log.

image

**Comment by membphis** [comment]
image

File: lib/resty/etcd.lua [hunk 2]
```diff
@@ -89,17 +106,19 @@ function _M.new(opts)
         return nil, 'opts must be table'
     end
 
-    local timeout = opts.timeout or 5000    -- 5 sec
+    local timeout = opts.timeout or 5    -- 5 sec
     local http_host = opts.host or "http://127.0.0.1:2379"
     local ttl = opts.ttl or -1
     local prefix = opts.prefix or "/v2/keys"
+    local user = opts.user or ""
```
Comment (membphis) [inline lib/resty/etcd.lua]: use empty string as default value is not a good idea. we have to check it by `user == ""`.

Therefore, if not specified, it is best to use `nil`, and we can check it by `if user then`.

**Comment by DoubleSpout** [comment]
I have complete cla test.
cluster unit test, i will write later

File: lib/resty/etcd.lua [hunk 3]
```diff
@@ -110,34 +129,104 @@ function _M.new(opts)
         return nil, 'opts.prefix must be string'
     end
 
+    if not typeof.string(user) then
```
Comment (membphis) [inline lib/resty/etcd.lua]: It should be an option, allowing unspecified

File: lib/resty/etcd.lua [hunk 4]
```diff
@@ -110,34 +129,104 @@ function _M.new(opts)
...
+
+local function choose_one_endpoint(self, endpoints)
+    if #endpoints == 0 then
+        return endpoints
+    end
+    local endpoints_len = #endpoints
+    if endpoints_len == 1 then
+        return endpoints[1]
+    end
+    self.init_count = (self.init_count or 0) + 1
```
Comment (membphis) [inline lib/resty/etcd.lua]: Add a blank line before it

File: lib/resty/etcd.lua [hunk 5]
```diff
@@ -110,34 +129,104 @@ function _M.new(opts)
...
+
+    return endpoints[pos]
+end
+
+
+
+-- return key, value
+-- example: 'Authorization', 'Basic dsfsfsddsfddsdsffd=='
+local function create_basicauth(user, password)
+    local userPwd = string.format('%s:%s', user, password)
```
Comment (membphis) [inline lib/resty/etcd.lua]: change to this style: `user .. ":" .. password`

**Comment by DoubleSpout** [comment]
add cluster unit test case~

File: lib/resty/etcd.lua (hunk 4 above)
Comment (DoubleSpout) [reply lib/resty/etcd.lua]: has added

**Comment by membphis** [comment]
@DoubleSpout we need to add a new test case to cover basic auth in v2 for etcd.

We can run this test case alone, eg:

```shell
# stop current etcd
# start new etcd 
# set basic auth
# run test case
```

**Comment by membphis** [comment]
@nic-chen please take a look when you have time ^_^

File: t/key.t [hunk 6]
```diff
@@ -305,3 +305,96 @@ GET /t
...
+                    "http://127.0.0.1:32379",
+                }
+            })
+            check_res(etcd, err)
+
+            local res, err = etcd:set("/test", {a = "abc"})
+            check_res(res, err)
+
+            ngx.sleep(1)
+
```
Comment (nic-chen) [inline t/key.t]: 建议在删除前读一次，以便验证读取OK

File: t/key.t [hunk 7]
```diff
@@ -305,3 +305,96 @@ GET /t
...
+    }
+--- request
+GET /t
+--- no_error_log
+[error]
+--- response_body
+res: nil err: opts.user must be string or ignore
+res: nil err: opts.password must be string or ignore
+--- timeout: 5
+
```
Comment (nic-chen) [inline t/key.t]: 需要一个正常验证Basic Auth的case

File: t/key.t (hunk 6 above)
Comment (moonming) [reply t/key.t]: @nic-chen please use English

**Comment by membphis** [comment]
merged, thx. will do the todo things later.

**Comment by DoubleSpout** [comment]
sorry, I'm busy recently, I will add basic auth test case lately(about this weekend), add new pr ^_^

**Comment by membphis** [comment]
good ^^

**Pull request merged**